[ClangImporter] Swift needs to pass `-Xclang -fbuiltin-headers-in-system-modules` for its module maps that group cstd headers

ian-twilightcoder opened the pull request:
<!-- What's in this pull request? -->
Swift has some module maps it overlays on Linux and Windows that groups all of the C standard library headers into a single module. This doesn’t allow clang and C++ headers to layer properly with the OS/SDK modules. clang will set -fbuiltin-headers-in-system-modules as necessary for Apple SDKs, but Swift will need to pass that flag itself when required by its module maps.

rdar://118361742

al45tair commented:
> Backtracing is currently just absorbing them all into its OS module

Indeed. Sadly that was necessary because of when this code gets built (it's before the `Darwin` or `Glibc` modules are available).

al45tair commented:
@swift-ci Please test

ian-twilightcoder commented:
I'm still testing this with https://github.com/apple/llvm-project/pull/7619, looks like it's not working yet.

ian-twilightcoder commented:
This turned out to be a bigger change, I'll clean up the PR tomorrow, I just pushed the branch to run the CI overnight.

ian-twilightcoder commented:
Running one more test in https://github.com/apple/llvm-project/pull/7619 and then I'll put this through CI

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci test Windows

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

glessard commented on stdlib/public/Backtracing/SymbolicatedBacktrace.swift:
```diff
@@ -23,6 +23,11 @@ import Swift
 @_implementationOnly import Runtime
+// Because we've turned off the OS/SDK modules, and we don't have a module for
+// stddef.h, and we sometimes build with -fbuiltin-headers-in-system-modules for
+// vfs reasons, stddef.h can be absorbed into a random module. Sometimes it's
+// SwiftOverlayShims.
```
> Sometimes it's SwiftOverlayShims.

This sounds stochastic. Is there a way to give it a stable home regardless of flags?

ian-twilightcoder replied:
Backtracing would need to stop disabling intrinsic module maps so that it could use the OS/SDK module maps where stddef.h lives. I'm not really sure why it does that. I guess Backtracing could add more comprehensive module maps to cover the actual headers instead of assuming that Libc.h will absorb all of the effectively non-modular OS/SDK headers into Backtracking's OS.Libc module.

al45tair replied:
It does it because the `Darwin` module (and its counterparts on Linux and Windows) are built later on in the build process than the backtracer, and attempting to fix that created dependency cycles in the CMake build system that were difficult to fix.

al45tair replied:
I'd rather it wasn't the cause of random behaviour though; if we can avoid this by fixing the `OS.Libc` module map then I'd be in favour of doing that instead.

ian-twilightcoder replied:
The Swift overlays get built later, but does that preclude us from using the clang modules? Fixing this in `OS.Libc` would amount to copying the module maps we vfs overlay on Linux/Windows, and copying the Apple module maps. I think it would be better to just use them.

But the behavior has always been random with the way Backtracing is set up, I would prefer to just add this import here to keep Linux limping along, and do a better fix later.

al45tair replied:
There is an additional problem that _mainly_ affects Darwin, namely that some of the definitions in the C headers are written in a manner that precludes their use from Swift, so I don't think it's _quite_ as simple as "just use the system module maps". I _did_ try that approach before deciding that it was better to do what it currently does. I have no objection to trying to use the system ones again, in principle, but it didn't work before.

ian-twilightcoder replied:
Seeing as how this isn't making the problem any worse, it's just landing on a different chaos number, could we possibly just go with this for now and try to fix it for real later?

al45tair replied:
Seems good to me.

ian-twilightcoder replied:
Filed rdar://119565504 (amongst the other ones I mentioned) to take a new look at this.

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@compnerd any chance you could help me with the Windows failures? I'm pretty sure the issue is that `-fbuiltin-headers-in-system-modules` isn't making it down to `ModuleMapParser::parseHeaderDecl` in llvm-project/clang/lib/Lex/ModuleMap.cpp, but I don't have a way to actually debug on Windows. I tried debugging on macOS as best I could using `-target x86_64-unknown-windows-msvc` and checking the `importer->Impl.Instance.getLangOpts().BuiltinHeadersInSystemModules` in `ClangImporter::create` in `swift/lib/ClangImporter/ClangImporter.cpp` and it's always `true` there. I suspect it's getting lost somewhere between `ClangImporter` and `ModuleMapParser`, but all my attempts to debug die trying to load the standard Swift library for Windows which I'm not sure how to get on a Mac since all the downloads on swift.org are .exe files.

compnerd commented:
> @compnerd any chance you could help me with the Windows failures? I'm pretty sure the issue is that `-fbuiltin-headers-in-system-modules` isn't making it down to `ModuleMapParser::parseHeaderDecl` in llvm-project/clang/lib/Lex/ModuleMap.cpp, but I don't have a way to actually debug on Windows.

I don't think that is the case at all if you look at the error logs.

> <unknown>:0: error: cannot open file 'C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\vcruntime.apinotes' (no such file or directory)

I think that the changes might not be honouring the VFS overlay as modifying the VS was undesirable, so this injection is now done through an injected VFS overlay by the frontend.

ian-twilightcoder commented:
> > @compnerd any chance you could help me with the Windows failures? I'm pretty sure the issue is that `-fbuiltin-headers-in-system-modules` isn't making it down to `ModuleMapParser::parseHeaderDecl` in llvm-project/clang/lib/Lex/ModuleMap.cpp, but I don't have a way to actually debug on Windows.
> 
> I don't think that is the case at all if you look at the error logs.
> 
> > :0: error: cannot open file 'C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\vcruntime.apinotes' (no such file or directory)
> 
> I think that the changes might not be honouring the VFS overlay as modifying the VS was undesirable, so this injection is now done through an injected VFS overlay by the frontend.

🤔 I only see that error in a few of the failed tests, is `-code-completion` eating that error in the other tests? Which way is the vfs supposed to be seen in those tests? Via the `-Xcc -ivfsoverlay` argument or injected by `swift::getClangInvocationFileMapping`? Why is `-Xcc -ivfsoverlay` passed instead of `-vfsoverlay`? I wonder if that was making the command line overlay be ignored until I started passing `-strict-implicit-module-context` which was needed for `-Xcc -fbuiltin-headers-in-system-modules` to make it through (otherwise `InterfaceSubContextDelegateImpl` in `ModuleInterfaceLoader.cpp` drops `ClangImporterOptions.ExtraArgs`). Maybe I should try taking out `-Xcc -ivfsoverlay` from the lit config since I _think_ it was probably being ignored anyway, but keep `-strict-implicit-module-context` where we use `-vfsoverlay`.

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
> > > @compnerd any chance you could help me with the Windows failures? I'm pretty sure the issue is that `-fbuiltin-headers-in-system-modules` isn't making it down to `ModuleMapParser::parseHeaderDecl` in llvm-project/clang/lib/Lex/ModuleMap.cpp, but I don't have a way to actually debug on Windows.
> > 
> > 
> > I don't think that is the case at all if you look at the error logs.
> > > :0: error: cannot open file 'C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\vcruntime.apinotes' (no such file or directory)
> > 
> > 
> > I think that the changes might not be honouring the VFS overlay as modifying the VS was undesirable, so this injection is now done through an injected VFS overlay by the frontend.
> 
> 🤔 I only see that error in a few of the failed tests, is `-code-completion` eating that error in the other tests? Which way is the vfs supposed to be seen in those tests? Via the `-Xcc -ivfsoverlay` argument or injected by `swift::getClangInvocationFileMapping`? Why is `-Xcc -ivfsoverlay` passed instead of `-vfsoverlay`? I wonder if that was making the command line overlay be ignored until I started passing `-strict-implicit-module-context` which was needed for `-Xcc -fbuiltin-headers-in-system-modules` to make it through (otherwise `InterfaceSubContextDelegateImpl` in `ModuleInterfaceLoader.cpp` drops `ClangImporterOptions.ExtraArgs`). Maybe I should try taking out `-Xcc -ivfsoverlay` from the lit config since I _think_ it was probably being ignored anyway, but keep `-strict-implicit-module-context` where we use `-vfsoverlay`.

Awesome that made the "can't find file" errors all go away and the missing string one too. There's just one last "rebuilt a thing we didn't think we'd need to rebuild" error which isn't nearly so bad. I'll rerun the tests and see if it goes away.

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci smoke test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci test

ian-twilightcoder commented:
> > > > @compnerd any chance you could help me with the Windows failures? I'm pretty sure the issue is that `-fbuiltin-headers-in-system-modules` isn't making it down to `ModuleMapParser::parseHeaderDecl` in llvm-project/clang/lib/Lex/ModuleMap.cpp, but I don't have a way to actually debug on Windows.
> > > 
> > > 
> > > I don't think that is the case at all if you look at the error logs.
> > > > :0: error: cannot open file 'C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\vcruntime.apinotes' (no such file or directory)
> > > 
> > > 
> > > I think that the changes might not be honouring the VFS overlay as modifying the VS was undesirable, so this injection is now done through an injected VFS overlay by the frontend.
> > 
> > 
> > 🤔 I only see that error in a few of the failed tests, is `-code-completion` eating that error in the other tests? Which way is the vfs supposed to be seen in those tests? Via the `-Xcc -ivfsoverlay` argument or injected by `swift::getClangInvocationFileMapping`? Why is `-Xcc -ivfsoverlay` passed instead of `-vfsoverlay`? I wonder if that was making the command line overlay be ignored until I started passing `-strict-implicit-module-context` which was needed for `-Xcc -fbuiltin-headers-in-system-modules` to make it through (otherwise `InterfaceSubContextDelegateImpl` in `ModuleInterfaceLoader.cpp` drops `ClangImporterOptions.ExtraArgs`). Maybe I should try taking out `-Xcc -ivfsoverlay` from the lit config since I _think_ it was probably being ignored anyway, but keep `-strict-implicit-module-context` where we use `-vfsoverlay`.
> 
> Awesome that made the "can't find file" errors all go away and the missing string one too. There's just one last "rebuilt a thing we didn't think we'd need to rebuild" error which isn't nearly so bad. I'll rerun the tests and see if it goes away.

I'm suspicious of the error I'm getting from `-emit-pcm`/`-verify` where it looks like the pcm is being emitted without `-fbuiltin-headers-in-system-modules`. I'm also having too much trouble trying to debug this without Windows, so I think what I'm going to do is try setting that flag in the MSVC driver/toolchain in clang for now. When I have a little breathing room, we'll either have removed that flag from Darwin and I'll have a better change at debugging how `-fbuiltin-headers-in-system-modules` flows through Swift, or we can fix the ucrt module to not require `-fbuiltin-headers-in-system-modules`, which we should do for all of our module maps anyway. Testing with the workaround.

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci smoke test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci smoke test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci smoke test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci smoke test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci smoke test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci smoke test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7848
@swift-ci test

ian-twilightcoder committed:
[ClangImporter] Swift needs to pass `-Xclang -fbuiltin-headers-in-system-modules` for its module maps that group cstd headers

Swift has some module maps it overlays on Linux and Windows that groups all of the C standard library headers into a single module. This doesn’t allow clang and C++ headers to layer properly with the OS/SDK modules. clang will set -fbuiltin-headers-in-system-modules as necessary for Apple SDKs, but Swift will need to pass that flag itself when required by its module maps.

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7619
@swift-ci smoke test

ian-twilightcoder commented:
https://github.com/apple/llvm-project/pull/7619
@swift-ci test

ian-twilightcoder commented:
@swift-ci smoke test

ian-twilightcoder commented:
@swift-ci test

al45tair commented on lib/ClangImporter/ClangIncludePaths.cpp:
```diff
@@ -515,20 +529,36 @@ ClangInvocationFileMapping swift::getClangInvocationFileMapping(
+  // SwiftLibc module too, the cycle goes away. Note that
+  // -fbuiltin-headers-in-system-modules does nothing to fix the same problem
+  // with C++ headers, and is generally fragile.
+  result.requiresBuiltinHeadersInSystemModules = !libcFileMapping.empty();
```
Presumably we're going to have some bug reports filed to fix these module maps? (The Musl one I've been working on is already split out because of this problem, so when I rebase on top of this I'll have to make sure to not set `requiresBuiltinHeadersInSystemModules` in that case.)

al45tair commented on test/DebugInfo/C-typedef-Darwin.swift:
```diff
@@ -9,4 +9,13 @@ let blah = size_t(1024)
+// Darwin.POSIX._types._size_t module. Ideally Apple will remove the duplicate
+// declaration and clang will settle to _Builtin_stddef.size_t, but while it's
+// all in flux allow all three of those modules.
+// Darwin.C.stddef|_Builtin_stddef.size_t|Darwin.POSIX._types._size_t
```
Likewise, I assume the relevant bug reports have been filed for this so that we actually tidy this up?

ian-twilightcoder replied:
rdar://101573150 is this specific one, there are some other similar ones, and probably a few more I need to find and file

al45tair reviewed (approved):
LGTM.

finagolfin commented on test/ClangImporter/pcm-emit-direct-cc1-mode.swift:
```diff
@@ -2,6 +2,9 @@
 // RUN: %swift-frontend -emit-pcm -direct-clang-cc1-module-build -only-use-extra-clang-opts -module-name script -o %t/script.pcm %S/Inputs/custom-modules/module.modulemap -Xcc %S/Inputs/custom-modules/module.modulemap -Xcc -o -Xcc %t/script.pcm -Xcc -fmodules -Xcc -triple -Xcc %target-triple -Xcc -x -Xcc objective-c -dump-clang-diagnostics 2> %t.diags.txt
 
+// Sometimes returns a 1 exit code with no stdout or stderr or even an indication of which command failed.
+// XFAIL: OS=linux-gnu
```
@ian-twilightcoder, this pull [broke this test on the Android CI too](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android/6131/console): would you please submit a followup pull modifying this to `// XFAIL: OS=linux-gnu, OS=linux-android, OS=linux-androideabi`?

ian-twilightcoder replied:
Sorry about that, https://github.com/apple/swift/pull/70523
